File: boj_codes/C++17/10974.cpp
// https://www.acmicpc.net/problem/10974
// First Written : 20210118
// Last Modified : 20210118

#include <cstdio>
#include <vector>

std::vector<int> numbers;
bool visited[8] = {false};

void perm(int depth, const int N){
    if(depth < N){
        for(int i = 0; i < N; i++){
            if(!visited[i]){
                numbers.push_back(i + 1);
                visited[i] = true;

                perm(depth + 1, N);
                
                visited[i] = false;
                numbers.pop_back();
            }
        }
    }

    else{
        for(int i = 0; i < N; i++)
            printf("%d ", numbers[i]);
        printf("\n");
    }

}

int main(){
    int N;
    scanf("%d", &N);
    perm(0, N);

    return 0;
}

File: boj_codes/Python3/10974.py
# https://www.acmicpc.net/problem/10974
# 20201008

import sys


"""
def perm(numbers, depth, n):
    if depth == n:
        print_numbers(numbers)
        return

    else:
        for i in range(depth, n):
            swap(numbers, i, depth)
            perm(numbers, depth + 1, n)
            swap(numbers, i, depth)
        return

def swap(numbers, i, j):
    if i == j:
        return
    temp = numbers[i]
    numbers[i] = numbers[j]
    numbers[j] = temp

    return

"""

def print_numbers(nums):
    for num in nums:
        print(num, end=" ")
    print()

def perm(numbers, depth, n):
    if depth < n:
        for i in range(n):
            if visited[i] == 0:
                numbers.append(i + 1)
                visited[i] = 1
                perm(numbers, depth + 1, n)
                visited[i] = 0
                numbers.pop()
    else:
        print_numbers(numbers)

N = int(sys.stdin.readline().rstrip())
numbers = []
visited= [0] * N

perm(numbers, 0, N)


